repose-aggregator/components/filters/openstack-identity-v3/src/main/scala/org/openrepose/filters/openstackidentityv3/OpenStackIdentityV3Handler.scala

```diff
@@ -26,225 +26,149 @@ import com.rackspace.httpdelegation.HttpDelegationManager
+        None
+      case Failure(e: IdentityServiceOverLimitException) =>
+        filterFailure = Some(FilterFailure(HttpServletResponse.SC_SERVICE_UNAVAILABLE, e.getMessage))
+        authServiceOverLimit(e, httpResponse)
+        None
+      case Failure(e) =>
+        filterFailure = Some(FilterFailure(HttpServletResponse.SC_INTERNAL_SERVER_ERROR, e.getMessage))
+        logger.error(e.getMessage)
+        None
+    }
 
-  override def handleRequest(request: HttpServletRequest, response: ReadableHttpServletResponse): FilterDirector = {
-    val filterDirector: FilterDirector = new FilterDirectorImpl()
+    // Attempt to check the project ID if configured to do so
+    if (filterFailure.isEmpty && !isProjectIdValid(wrappedHttpRequest.getRequestURI, token.get)) {
+      filterFailure = Some(FilterFailure(HttpServletResponse.SC_UNAUTHORIZED, "Invalid project ID for token: " + token.get))
+    }
 
-    def delegateOrElse(responseCode: Int, message: String)(f: => Unit) {
-      delegatingWithQuality match {
-        case Some(quality) =>
-          buildDelegationHeaders(responseCode, "openstack-identity-v3", message, quality) foreach { case (key, values) =>
-            filterDirector.requestHeaderManager.appendHeader(key, values: _*)
-            filterDirector.setFilterAction(FilterAction.PROCESS_RESPONSE)
-            filterDirector.setResponseStatusCode(200) // Note: The response status code must be set to a non-500 so that the request will be routed appropriately.
-          }
-        case None =>
-          f
-      }
+    // Attempt to authorize the token against a configured endpoint
+    if (filterFailure.isEmpty && !isAuthorized(token.get)) {
+      filterFailure = Some(FilterFailure(HttpServletResponse.SC_FORBIDDEN, "Invalid endpoints for token: " + token.get))
     }
 
-    def authServiceOverLimit(e: IdentityServiceOverLimitException, filterDirector: FilterDirector): Unit = {
-      filterDirector.setResponseStatusCode(HttpServletResponse.SC_SERVICE_UNAVAILABLE)
-      var retry = e.getRetryAfter
-      if (retry == null) {
-        val retryCalendar = new GregorianCalendar
-        retryCalendar.add(Calendar.SECOND, 5)
-        retry = new HttpDate(retryCalendar.getTime).toRFC1123
+    // Attempt to fetch groups if configured to do so
+    val userGroups = if (filterFailure.isEmpty && filterConfig.fullConfig.isForwardGroups) {
+      token.get.user.id map { userId =>
+        identityApi.getGroups(userId, tracingHeader) match {
+          case Success(groupsList) =>
+            groupsList.map(_.name)
+          case Failure(e: IdentityServiceOverLimitException) =>
+            filterFailure = Some(FilterFailure(HttpServletResponse.SC_SERVICE_UNAVAILABLE, e.getMessage))
+            authServiceOverLimit(e, httpResponse)
+            List.empty[String]
+          case Failure(e) =>
+            filterFailure = Some(FilterFailure(HttpServletResponse.SC_INTERNAL_SERVER_ERROR, e.getMessage))
+            logger.error(e.getMessage)
```

Comment:
In some cases where Exceptions are expected (i.e., they are used as to control the flow) I understand only printing the messages and not the full stack trace.

Since this is a catch-all, however, I agree that printing the full stack trace is probably the right call.